doc/prebuilt/man/rtcansend.1
'\" t
.\"     Title: rtcansend
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2008/04/19
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 2.99.0
.\"  Language: English
.\"
.TH "RTCANSEND" "1" "2008/04/19" "Xenomai 2\&.99\&.0" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
rtcansend \- Xenomai tool for sending CAN messages
.SH "SYNOPSIS"
.sp
\fBrtcansend\fR <can\-interface> [Options] <can\-msg>
.SH "DESCRIPTION"
.sp
\fBrtcansend\fR is part of Xenomai\&. It is used to send messages via a CAN interface\&. The system must run a suitable Xenomai enabled kernel with the respective module (xeno_native and the CAN driver)\&.
.SH "OPTIONS"
.sp
<can\-interface> is the CAN interface file\&.
.sp
<can\-msg> can consist of up to 8 bytes given as a space separated list\&.
.sp
\fBrtcansend\fR accepts the following options:
.PP
\fB\-i, \-\-identifier=ID\fR
.RS 4
CAN Identifier (default = 1)
.RE
.PP
\fB\-r, \-\-rtr\fR
.RS 4
send remote request
.RE
.PP
\fB\-e, \-\-extended\fR
.RS 4
send extended frame
.RE
.PP
\fB\-l, \-\-loop=COUNT\fR
.RS 4
send message COUNT times
.RE
.PP
\fB\-c, \-\-count\fR
.RS 4
message count in data[0\-3]
.RE
.PP
\fB\-d, \-\-delay=MS\fR
.RS 4
delay in ms (default = 1ms)
.RE
.PP
\fB\-s, \-\-send\fR
.RS 4
use send instead of sendto
.RE
.PP
\fB\-t, \-\-timeout=MS\fR
.RS 4
timeout in ms
.RE
.PP
\fB\-L, \-\-loopback=0|1\fR
.RS 4
switch local loopback off or on
.RE
.PP
\fB\-v, \-\-verbose\fR
.RS 4
be verbose
.RE
.PP
\fB\-p, \-\-print=MODULO\fR
.RS 4
print every MODULO message
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
a usage description
.RE
.SH "SEE ALSO"
.sp
\fBrtcanconfig(1)\fR\fB, \fR\fBrtcanrecv(1)\fR\&.
.SH "AUTHORS"
.sp
\fBrtcansend\fR was written by Wolfgang Grandegger, Jan Kiszka and Philippe Gerum\&. This man page was written by Roland Stigge\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

utils/can/rtcansend.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <getopt.h>
#include <sys/mman.h>

#include <alchemy/task.h>
#include <alchemy/timer.h>

#include <rtdm/can.h>

extern int optind, opterr, optopt;

static void print_usage(char *prg)
{
    fprintf(stderr,
	    "Usage: %s <can-interface> [Options] <can-msg>\n"
	    "<can-msg> can consist of up to 8 bytes given as a space separated list\n"
	    "Options:\n"
	    " -i, --identifier=ID   CAN Identifier (default = 1)\n"
	    " -r  --rtr             send remote request\n"
	    " -e  --extended        send extended frame\n"
	    " -l  --loop=COUNT      send message COUNT times\n"
	    " -c, --count           message count in data[0-3]\n"
	    " -d, --delay=MS        delay in ms (default = 1ms)\n"
	    " -s, --send            use send instead of sendto\n"
	    " -t, --timeout=MS      timeout in ms\n"
	    " -L, --loopback=0|1    switch local loopback off or on\n"
	    " -v, --verbose         be verbose\n"
	    " -p, --print=MODULO    print every MODULO message\n"
	    " -h, --help            this help\n",
	    prg);
}


RT_TASK rt_task_desc;

static int s=-1, dlc=0, rtr=0, extended=0, verbose=0, loops=1;
static SRTIME delay=1000000;
static int count=0, print=1, use_send=0, loopback=-1;
static nanosecs_rel_t timeout = 0;
static struct can_frame frame;
static struct sockaddr_can to_addr;


static void cleanup(void)
{
    int ret;

    if (verbose)
	printf("Cleaning up...\n");

    usleep(100000);

    if (s >= 0) {
	ret = rt_dev_close(s);
	s = -1;
	if (ret) {
	    fprintf(stderr, "rt_dev_close: %s\n", strerror(-ret));
	}
	exit(EXIT_SUCCESS);
    }
}

static void cleanup_and_exit(int sig)
{
    if (verbose)
	printf("Signal %d received\n", sig);
    cleanup();
    exit(0);
}

static void rt_task(void)
{
    int i, j, ret;

    for (i = 0; i < loops; i++) {
	rt_task_sleep(rt_timer_ns2ticks(delay));
	if (count)
	    memcpy(&frame.data[0], &i, sizeof(i));
	/* Note: sendto avoids the definiton of a receive filter list */
	if (use_send)
	    ret = rt_dev_send(s, (void *)&frame, sizeof(can_frame_t), 0);
	else
	    ret = rt_dev_sendto(s, (void *)&frame, sizeof(can_frame_t), 0,
				(struct sockaddr *)&to_addr, sizeof(to_addr));
	if (ret < 0) {
	    switch (ret) {
	    case -ETIMEDOUT:
		if (verbose)
		    printf("rt_dev_send(to): timed out");
		break;
	    case -EBADF:
		if (verbose)
		    printf("rt_dev_send(to): aborted because socket was closed");
		break;
	    default:
		fprintf(stderr, "rt_dev_send: %s\n", strerror(-ret));
		break;
	    }
	    i = loops;		/* abort */
	    break;
	}
	if (verbose && (i % print) == 0) {
	    if (frame.can_id & CAN_EFF_FLAG)
		printf("<0x%08x>", frame.can_id & CAN_EFF_MASK);
	    else
		printf("<0x%03x>", frame.can_id & CAN_SFF_MASK);
	    printf(" [%d]", frame.can_dlc);
	    for (j = 0; j < frame.can_dlc; j++) {
		printf(" %02x", frame.data[j]);
	    }
	    printf("\n");
	}
    }
}

int main(int argc, char **argv)
{
    int i, opt, ret;
    struct ifreq ifr;
    char name[32];

    struct option long_options[] = {
	{ "help", no_argument, 0, 'h' },
	{ "identifier", required_argument, 0, 'i'},
	{ "rtr", no_argument, 0, 'r'},
	{ "extended", no_argument, 0, 'e'},
	{ "verbose", no_argument, 0, 'v'},
	{ "count", no_argument, 0, 'c'},
	{ "print", required_argument, 0, 'p'},
	{ "loop", required_argument, 0, 'l'},
	{ "delay", required_argument, 0, 'd'},
	{ "send", no_argument, 0, 's'},
	{ "timeout", required_argument, 0, 't'},
	{ "loopback", required_argument, 0, 'L'},
	{ 0, 0, 0, 0},
    };

    mlockall(MCL_CURRENT | MCL_FUTURE);

    signal(SIGTERM, cleanup_and_exit);
    signal(SIGINT, cleanup_and_exit);

    frame.can_id = 1;

    while ((opt = getopt_long(argc, argv, "hvi:l:red:t:cp:sL:",
			      long_options, NULL)) != -1) {
	switch (opt) {
	case 'h':
	    print_usage(argv[0]);
	    exit(0);

	case 'p':
	    print = strtoul(optarg, NULL, 0);

	case 'v':
	    verbose = 1;
	    break;

	case 'c':
	    count = 1;
	    break;

	case 'l':
	    loops = strtoul(optarg, NULL, 0);
	    break;

	case 'i':
	    frame.can_id = strtoul(optarg, NULL, 0);
	    break;

	case 'r':
	    rtr = 1;
	    break;

	case 'e':
	    extended = 1;
	    break;

	case 'd':
	    delay = strtoul(optarg, NULL, 0) * 1000000LL;
	    break;

	case 's':
	    use_send = 1;
	    break;

	case 't':
	    timeout = strtoul(optarg, NULL, 0) * 1000000LL;
	    break;

	case 'L':
	    loopback = strtoul(optarg, NULL, 0);
	    break;

	default:
	    fprintf(stderr, "Unknown option %c\n", opt);
	    break;
	}
    }

    if (optind == argc) {
	print_usage(argv[0]);
	exit(0);
    }

    if (argv[optind] == NULL) {
	fprintf(stderr, "No Interface supplied\n");
	exit(-1);
    }

    if (verbose)
	printf("interface %s\n", argv[optind]);

    ret = rt_dev_socket(PF_CAN, SOCK_RAW, CAN_RAW);
    if (ret < 0) {
	fprintf(stderr, "rt_dev_socket: %s\n", strerror(-ret));
	return -1;
    }
    s = ret;

    if (loopback >= 0) {
	ret = rt_dev_setsockopt(s, SOL_CAN_RAW, CAN_RAW_LOOPBACK,
				&loopback, sizeof(loopback));
	if (ret < 0) {
	    fprintf(stderr, "rt_dev_setsockopt: %s\n", strerror(-ret));
	    goto failure;
	}
	if (verbose)
	    printf("Using loopback=%d\n", loopback);
    }

    strncpy(ifr.ifr_name, argv[optind], IFNAMSIZ);
    if (verbose)
	printf("s=%d, ifr_name=%s\n", s, ifr.ifr_name);

    ret = rt_dev_ioctl(s, SIOCGIFINDEX, &ifr);
    if (ret < 0) {
	fprintf(stderr, "rt_dev_ioctl: %s\n", strerror(-ret));
	goto failure;
    }

    memset(&to_addr, 0, sizeof(to_addr));
    to_addr.can_ifindex = ifr.ifr_ifindex;
    to_addr.can_family = AF_CAN;
    if (use_send) {
	/* Suppress definiton of a default receive filter list */
	ret = rt_dev_setsockopt(s, SOL_CAN_RAW, CAN_RAW_FILTER, NULL, 0);
	if (ret < 0) {
	    fprintf(stderr, "rt_dev_setsockopt: %s\n", strerror(-ret));
	    goto failure;
	}

	ret = rt_dev_bind(s, (struct sockaddr *)&to_addr, sizeof(to_addr));
	if (ret < 0) {
	    fprintf(stderr, "rt_dev_bind: %s\n", strerror(-ret));
	    goto failure;
	}
    }

    if (count)
	frame.can_dlc = sizeof(int);
    else {
	for (i = optind + 1; i < argc; i++) {
	    frame.data[dlc] = strtoul(argv[i], NULL, 0);
	    dlc++;
	    if( dlc == 8 )
		break;
	}
	frame.can_dlc = dlc;
    }

    if (rtr)
	frame.can_id |= CAN_RTR_FLAG;

    if (extended)
	frame.can_id |= CAN_EFF_FLAG;

    if (timeout) {
	if (verbose)
	    printf("Timeout: %lld ns\n", (long long)timeout);
	ret = rt_dev_ioctl(s, RTCAN_RTIOC_SND_TIMEOUT, &timeout);
	if (ret) {
	    fprintf(stderr, "rt_dev_ioctl SND_TIMEOUT: %s\n", strerror(-ret));
	    goto failure;
	}
    }

    snprintf(name, sizeof(name), "rtcansend-%d", getpid());
    ret = rt_task_shadow(&rt_task_desc, name, 1, 0);
    if (ret) {
	fprintf(stderr, "rt_task_shadow: %s\n", strerror(-ret));
	goto failure;
    }

    rt_task();

    cleanup();
    return 0;

 failure:
    cleanup();
    return -1;
}
